### PR TITLE
feat(fr) StreamingDataset: Fault Tolerance v2 1/n

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces the `CombinedStreamingDataset` and the `StreamingDataLoader`. Both objects are responsible for tracking the number of samples fetched and return the associated state_dict when selected.

In another PR, I will re-write the checkpointing logic of the StreamingDataset with this new paradigm.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19196.org.readthedocs.build/en/19196/

cc @borda

### CODE REVIEW

File: src/lightning/data/streaming/combined.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        if len(state_dict) != len(self._datasets):
+            raise RuntimeError(
+                f"The provided state {state_dict} doesn't match the current number of datasets: {self._datasets}."
+            )
+
+        for dataset_idx, dataset in enumerate(self._datasets):
+            if str(dataset_idx) not in state_dict:
+                raise RuntimeError(f"The provided state {state_dict} doesn't contain the index {dataset_idx}.")
```
Comment (awaelchli) [inline src/lightning/data/streaming/combined.py:65]: I don't think we should print the entire state dict here, this will make it really hard to read the error message otherwise.
```suggestion
                raise RuntimeError(f"The provided state doesn't contain the index {dataset_idx}.")
```

File: src/lightning/data/streaming/combined.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+import random
+from typing import Any, Dict, Iterator, List, Optional, Sequence
+
+from torch.utils.data import IterableDataset
+
+from lightning.data.streaming.dataset import StreamingDataset
+
+
+class CombinedStreamingDataset(IterableDataset):
```
Comment (awaelchli) [inline src/lightning/data/streaming/combined.py:22]: Can you bring over the tests from lit-gpt?

File: src/lightning/data/streaming/combined.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    def __init__(
+        self, datasets: List[StreamingDataset], seed: int = 42, weights: Optional[Sequence[float]] = None
+    ) -> None:
+        self._seed = seed
+        self._datasets = datasets
+        self._weights = weights
+        num_datasets = len(datasets)
+
+        # Inversely weighted based on length
```
Comment (awaelchli) [inline src/lightning/data/streaming/combined.py:39]: This comment only applies to the first case, can you move it inside?

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/streaming/combined.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/combined.py:65]: Sure.

File: src/lightning/data/streaming/combined.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/combined.py:39]: sure.

File: src/lightning/data/streaming/combined.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/streaming/combined.py:22]: done !

**Review by dmitsf — APPROVED** [review]

**Comment by Borda** [comment]
@awaelchli @tchaton could you pls revisit it? It seems that after unfreezing requirements (so any PL user installing our package), all tests are failing... https://github.com/Lightning-AI/pytorch-lightning/commit/c989a97aa196431ecb70e1a0f8c839cdeaba9300

**Comment by tchaton** [comment]
@Borda Where are the failuresm can you share a link  ?

**Comment by Borda** [comment]
> Where are the failures, can you share a link ?

https://github.com/Lightning-AI/pytorch-lightning/actions/runs/7291059217
